azure/shared/client.js

```diff
@@ -0,0 +1,151 @@
+const STRING_ARRAY = 'string-array'; // example: ['one message', 'two message', ...]
+const JSON_OBJECT = 'json-object'; // example: {"key": "value"}
+const JSON_RECORDS = 'json-records'; // example: [{"records": [{}, {}, ...]}, {"records": [{}, {}, ...]}, ...]
+const JSON_ARRAY = 'json-array'; // example: [{"key": "value"}, {"key": "value"}, ...]
+const INVALID = 'invalid';
+
+var DD_API_KEY = process.env.DD_API_KEY || '<DATADOG_API_KEY>';
+var DD_URL = process.env.DD_URL || 'functions-intake.logs.datadoghq.com';
+var DD_TAGS = process.env.DD_TAGS || '<TAG_KEY>:<TAG_VALUE>';
+var DD_PORT = process.env.DD_PORT || 10516;
+var DD_SERVICE = process.env.DD_SERVICE || 'azure';
+var DD_SOURCE = process.env.DD_SOURCE || 'azure';
+var DD_SOURCE_CATEGORY = process.env.DD_SOURCE_CATEGORY || 'azure';
+
+function getSocket(context) {
+    var socket = tls.connect({ port: DD_PORT, host: DD_URL });
+    socket.on('error', err => {
+        context.log.error(err.toString());
+        socket.end();
+    });
+
+    return socket;
+}
+
+function send(socket, record) {
+    return socket.write(DD_API_KEY + ' ' + JSON.stringify(record) + '\n');
+}
+
+function handleLogs(handler, logs, context) {
+    var logsType = getLogFormat(logs);
+    switch (logsType) {
+        case STRING:
+            handler(addTagsToStringLog)(logs);
+            break;
+        case JSON_OBJECT:
+            handler(addTagsToJsonLog)(logs);
+            break;
+        case STRING_ARRAY:
+            logs.forEach(handler(addTagsToStringLog));
+            break;
+        case JSON_RECORDS:
+            logs.forEach(message => {
+                message.records.forEach(handler(addTagsToJsonLog));
+            });
+            break;
+        case JSON_ARRAY:
+            logs.forEach(handler(addTagsToJsonLog));
+            break;
+        case INVALID:
+        default:
+            context.log.warn('logs format is invalid');
+            break;
+    }
+}
+
+function getLogFormat(logs) {
+    if (typeof logs === 'string') {
+        return STRING;
+    }
+    if (!Array.isArray(logs) && typeof logs === 'object' && logs !== null) {
+        return JSON_OBJECT;
+    }
+    if (!Array.isArray(logs)) {
+        return INVALID;
+    }
+    if (typeof logs[0] === 'object') {
+        if (logs[0].records !== undefined) {
+            return JSON_RECORDS;
+        } else {
+            return JSON_ARRAY;
+        }
+    }
+    if (typeof logs[0] === 'string') {
+        return STRING_ARRAY;
+    }
+    return INVALID;
+}
+
+function addTagsToJsonLog(record, context) {
+    metadata = extractResourceId(record);
+    record['ddsource'] = metadata.source || DD_SOURCE;
+    record['ddsourcecategory'] = DD_SOURCE_CATEGORY;
+    record['service'] = DD_SERVICE;
+    record['ddtags'] = metadata.tags
+        .concat([
+            DD_TAGS,
+            'forwardername:' + context.executionContext.functionName
+        ])
+        .filter(Boolean)
+        .join(',');
+    return record;
```

Comment:
function that modifies first param & returns it... kind of a code smell.

Reply:
Yes, because this is one of the `tagger` function. If you call this one directly, you could reuse the 1st argument, but if call the one below `addTagsToStringLog`, you cannot because you pass a string and gets a new object. 

I'd be interested if you have a suggestion, ideally which doesn't allocate a new object if we call `addTagsToJsonLog` directly